### PR TITLE
feat: jupyter playground

### PR DESCRIPTION
**Pull request by jochumb**
This PR is a follow-up on this [slack message](https://devlake-io.slack.com/archives/C03ASGL114N/p1707844783669659), referencing this [article](https://xebia.com/blog/insights-from-your-jira-data-to-help-improve-your-team/).

The original goal was to make available the process graph shown in the article, to the community of DevLake users.
However, this couldn't be achieved within the current capabilities of DevLake, as it doesn't qualify as a plugin, and it requires `graphviz` as a container-level dependency and a way to provide the data to Grafana and show it in Grafana.
We concluded that the specific visualization would be just one example of a more flexible way to explore the data.
This PR is our idea of how this could be achieved, a "data playground" using the strengths of `python` with `pandas` in `jupyter`.

In addition to the Graphviz graph, we added a boxplot visualization on the same data to show duration statistics for the most occurring status transitions using `plotly`.

## Potential next step

Using this Jupyter playground for data exploration requires a clone of the repo and a local development setup.
It is also possible to add a Jupyter server as another (optional) container to the docker-compose set-up.
This way the notebooks could be used in a browser.

## Not yet included

- Running the tests in a GitHub action.
- Postgres support (no dependency on a client or connector has been added.)

### CODE REVIEW

**Comment by d4x1** [comment]
It's a plugin or a feature? How can we use it ?
I think some addtional documents are needed.

**Comment by jochumb** [comment]
> It's a plugin or a feature? How can we use it ? I think some addtional documents are needed.

@d4x1 We updated the PR with a description and also documentation within the change. I hope this clarifies our goals!

**Comment by lenntt** [comment]
Adding: proposing a change to easily filter on issue type, dates and project key.

**Comment by lenntt** [comment]
image
Added the following (pairing with @jochumb), so that a user can change whether to see the avg, mean, IQR or minmax as default, showing them all in the tooltip.

**Comment by d4x1** [comment]
Give me some time, and I'll review this PR.

File: playground/tests/process_analysis/status_transition_graph_test.py
Comment (d4x1) [inline playground/tests/process_analysis/status_transition_graph_test.py:1]: No need to follow golang stype test. Just use Python style, I am not sure which ("test_xxx" or "xxx_test") will be better, just follow Python style. (I have sawn "test_xxx" more frequently.)

File: playground/playground/process_analysis/issue_filter.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import pandas as pd
+
+@dataclass
+class IssueFilter:
+    project_key: Optional[str] = None
+    from_date: Optional[pd.Timestamp] = None
+    to_date: Optional[pd.Timestamp] = None
+    issue_type: Optional[str] = None
+
+    def apply(self, issue_df: pd.DataFrame ):
```
Comment (d4x1) [inline playground/playground/process_analysis/issue_filter.py:13]: Try to format code with a formatter.

File: playground/playground/process_analysis/status_transition_graph.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    created_date: pd.Timestamp
+    original_from_value: str
+    from_value: str
+    original_to_value: str
+    to_value: str
+    changed_date: pd.Timestamp
+
+
+class StatusTransitionGraph:
+    """A directed graph that represents the transitions between statuses of issues from a DevLake data source."""
```
Comment (d4x1) [inline playground/playground/process_analysis/status_transition_graph.py:24]: Maybe 
```
"""
doc string
"""
```
will be more pythonic?

**Comment by d4x1** [comment]
1. Here is a new repo  https://github.com/apache/incubator-devlake-playground,  it's created for this playground. You can remove codes to this repo.
DevLake repo is becoming more and more complicated and playground is an independent part.  With a standalone repo, it can be updated conveniently.(For example, devlake is still using Python3.9, which is outdated.)
2. I haven't run your code locally, but I have reviewed them, you can see the comments on codes.

**Comment by lenntt** [comment]
> 1. Here is a new repo  https://github.com/apache/incubator-devlake-playground,  it's created for this playground. You can remove codes to this repo.
>    DevLake repo is becoming more and more complicated and playground is an independent part.  With a standalone repo, it can be updated conveniently.(For example, devlake is still using Python3.9, which is outdated.)
> 2. I haven't run your code locally, but I have reviewed them, you can see the comments on codes.
> 
> Thanks for your contribution. I think adding jupyter playground deserves a new blog post in DevLake's official website.

thanks! I think moving to a seperate repository makes sense (need to keep in mind there might be some hard coupling by the data model - when it comes to releasing and versioning perhaps).


The repo is empty I think it needs a first commit because we can't fork it or open a PR:
image

**Comment by d4x1** [comment]
@lenntt https://github.com/apache/incubator-devlake-playground is not empty now.(I haven't noticed that empty repos cannot be forked.)

**Pull request closed without merging**